### ISSUE TITLE
NameError: uninitialized constant Foo::Bar Did you mean? Foo::Bar

### ISSUE THREAD

**Issue by ojab** [fxn/zeitwerk#100]
I'm trying to port rather large rails application to zeitwerk & have a strange issue. Unfortunately I can't construct reproducer or provide actual code, but I'll try to explain it as comprehensive as I could.

Basically I have `require 'analytics/consistency/local_schema/association_reflection_builder'` in spec & while loading that file with zeitwerk enabled it fails with
```
An error occurred while loading ./spec/lib/analytics/consistency/local_schema/association_reflection_builder_spec.rb.
Failure/Error: extend Helpers

NameError:
  uninitialized constant Analytics::Consistency::LocalSchema::Helpers
  Did you mean?  Analytics::Consistency::LocalSchema::Helpers
```

Simplified files look like that:
- lib/analytics/consistecy.rb
```ruby
module Analytics
  module Consistency
  end
end

require 'analytics/consistency/local_schema'
```

- lib/analytics/consistency/local_schema.rb
```ruby
require 'analytics/consistency/local_schema/helpers'
require 'analytics/consistency/local_schema/association_reflection_builder'

module Analytics
  module Consistency 
    module LocalSchema
      extend Helpers

      # AssociationReflectionBuilder usage
    end
  end
end
```

- lib/analytics/consistency/local_schema/association_reflection_builder.rb
```ruby
require 'analytics/consistency/local_schema/helpers'

module Analytics
  module Consistency
    module LocalSchema
      class AssociationReflectionBuilder
        include Helpers
      end
    end
  end
end
```

Helpers is just a module without any requires.

Related excerpt from autoloaders log: https://gist.github.com/ojab/91b1a025fd10a5dde7232bccd880dcc3

If I copy whole `/lib` folder to the new rails project — everything works fine.

So the question is: how can I fix it?

**Comment by ojab** [comment]
Forgot to mention: it can be workarounded by `require 'lib/analytics/consistency/local_schema'` instead of `require 'analytics/consistency/local_schema/association_reflection_builder'`

**Comment by ojab** [comment]
And the issue it seems in `config.paths['lib'].autoload!`. Reproducer: https://github.com/ojab/zeitwerk_testapp
```
$ bundle exec rails c
Loading development environment (Rails 6.0.1)
2.6.5 :001 > require 'foo/bar'
Traceback (most recent call last):
        5: from (irb):1
        4: from lib/foo/bar.rb:1:in `<main>'
        3: from lib/foo/helpers.rb:1:in `<main>'
        2: from lib/foo.rb:4:in `<main>'
        1: from lib/foo.rb:5:in `<module:Foo>'
NameError (uninitialized constant Foo::Helpers)
2.6.5 :002 > Foo.constants
 => [:Bar, :Helpers]
2.6.5 :003 >
```

**Comment by fxn** [comment]
Files within a directory managed by Zeitwerk should not have `require` calls, just remove them.

**Comment by ojab** [comment]
So it's basically expected behavior?

**Comment by fxn** [comment]
Yes, it is expected, I explain the reasoning below.

Firstly, though, let me explain that Zeitwerk is designed for autoloading. Files in the autoload paths shouldn't `require` each other like that. Same way you do not `require "user"` in `UsersController`. You always refer to the constants, and they are autoloaded on demand.

Zeitwerk has no public contract about how `require` works. It has undocumented support for reasonable use cases of `require`, but not any imaginable combination.

In your case, it goes like this:

1. Since `lib` is in the autoload paths, when the application boots Zeitwerk sets an autoload for `Foo` that autoloads `lib/foo.rb`.
2. The script issues a `require "foo/bar"`.
3. At the top of that file, there is a `require "foo/helper"'.
4. Line `module Foo` is reached, so Ruby executes the autoload configured in step 1.
5. When evaluating `lib/foo.rb` due to the autoload, the `require` calls at the top return `false` because those two files are in the stack of files being required.
6. Finally, you reach `module Foo`, which defines `Foo`.
7. Next line is `extend Helpers`, but if you follow the steps above, we did not reach the definition of `Foo::Helpers` because we jumped in step 4 right before the spot that defines it.

Bottom line: Remove all the `require` calls :).

**Comment by ojab** [comment]
Thanks! It's a shame to waste round issue number like that :)

**Comment by fxn** [comment]
Haha, all issues are important :).

Enjoy your `require` freedom!
